### ISSUE TITLE
NodeJS Critical API Difference (between v0.8 and v0.10)

### ISSUE THREAD

**Issue by coaster3000** [gruntjs/grunt#1321]
## Issue
There is a api change within nodejs that causes issues with multiple commands within grunt, however so far only two of them show issues unless you are running other plugins, then some of their commands also fail.

*Note: The issue does occur on bare minimal project GruntFile's. An error would occur if you left the file blank, stating that a gruntfile doesn't exist.*

[Wiki page containing list of api changes](https://github.com/joyent/node/wiki/Api-changes-between-v0.8-and-v0.10)

### Commands Affected
- grunt --tasks
- grunt --npm

Other commands are not part of grunt directly, but via plugins. Those ones are not included in here as there could be any number of plugins out there that have the issue.

**This issue affects all versions of nodejs above Version 0.8**

## Version Information

**Grunt Version Output**
grunt-clii v0.1.13
grunt v0.4.5

**Node Version Output**
v0.12.2

**npm version output**
2.7.4

### Command Output
**grunt --npm**
```
path.js:233
      throw new TypeError('Arguments to path.join must be strings');
            ^
TypeError: Arguments to path.join must be strings
    at Object.win32.join (path.js:233:13)
    at task.loadNpmTasks ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\task.js:383:22)
    at Array.forEach (native)
    at Task.task.init [PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\task.js:455:31)
    at Object.grunt.tasks ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt.js:120:8)
    at Object.module.exports [as cli] ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\cli.js:38:9)
    at Object.<anonymous> ([PathRemoved]\Roaming\npm\node_modules\grunt-cli\bin\grunt:45:20)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
```

**grunt --tasks**

```
path.js:233
      throw new TypeError('Arguments to path.join must be strings');
            ^
TypeError: Arguments to path.join must be strings
    at Object.win32.join (path.js:233:13)
    at Object.file.exists ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\file.js:381:28)
    at task.loadTasks ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\task.js:371:18)
    at Array.forEach (native)
    at Task.task.init ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\task.js:457:33)
    at Object.grunt.tasks ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt.js:120:8)
    at Object.module.exports [as cli] ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\cli.js:38:9)
    at Object.<anonymous> ([PathRemoved]\npm\node_modules\grunt-cli\bin\grunt:45:20)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
```

**Comment by vladikoff** [comment]
@cowboy  ^

**Comment by coaster3000** [comment]
Heh, I was sure I issued the command in a bare minimal environment for this issue, I will have to edit that in somewhere at some point when I get time.

I do know that it will still have the error.

**Comment by cowboy** [comment]
I thought we had fixed this issue in 5a1a8f8 and c6f83ca (and maybe others).

I see these lines in your stack traces:

* `at task.loadNpmTasks ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\task.js:383:22)`
* `at task.loadTasks ([PathRemoved]\workspace\node.js\grunt-bower-task\node_modules\grunt\lib\grunt\task.js:371:18)`

So, the [grunt-bower-task](https://github.com/yatskevich/grunt-bower-task) plugin probably has something to do with this issue.

Does the stack trace mean the Grunt inside `grunt-bower-task\node_modules\grunt` is being used to load tasks? If so, is that Grunt version `0.4.5` or an older version?

Is it possible that the grunt-bower-task plugin is using `path.join` incorrectly? I see these uses in their code:

```
grunt-bower-task/tasks/lib/layouts_manager.js
15:    return path.join(type, pkg);
19:    return path.join(pkg, type);

grunt-bower-task/tasks/lib/bower_assets.js
14:  var pkgPath = path.join(this._componentsDir, pkg);
32:  var basePath = path.join(this._cwd, pkgPath);
36:    return path.join(pkgPath, expandedPath);
57:  var bowerConfig = grunt.file.readJSON(path.join(this.cwd, this.config));

grunt-bower-task/tasks/lib/asset_copier.js
43:      var destinationDir = path.join(self.options.targetDir, self.options.layout(type, pkg, source));
46:        destination = path.join(destinationDir, path.basename(source));
```

**Comment by vladikoff** [comment]
Actually running `grunt --npm` or `--tasks` will give you that warning in all versions of node. I think this is an old issue

**Comment by coaster3000** [comment]
Yea cowbow, It could be grunt-bower-task also, However at some point I did test a bare minimal setup of grunt, and the issue did occur. I just ran the commands in wrong workspace when I added the output of the commands for the ticket. Sorry about that.

**Comment by xiaoshao** [comment]
I have the same problem.

**Comment by xiaoshao** [comment]
In the issue https://github.com/gruntjs/grunt/issues/726#issuecomment-19949327 , They said that they have fixed this problem in `grunt-bower-task 0.4.0`. 

And in my project, I use grunt-bower-task 0.4.0. but the error still occurs.
